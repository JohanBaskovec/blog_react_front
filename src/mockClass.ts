

type constructorTypeOf<T> = new (...args:any[]) => T;

export function mockDefaultApi<T>(o: T): T {
    const mocked: any = {};
    for (const key in o) {
        if (typeof o[key] === 'function') {
            mocked[key] = jest.fn();
        }
    }
    return mocked;
}

type MockedProxy<T> = jest.Mock<T extends (...args: any[]) => infer U ? U : T, any[]>;

type MockedClass<T> = {
    [k in keyof T]: MockedProxy<T[k]>;
}

export function mockClass<T>(constr: constructorTypeOf<T>): T & MockedClass<T>{
    let obj: T = new constr();
    const mocked: any = {};

    do {
        const properties = Object.getOwnPropertyNames(obj);
        for (const propertyName of properties) {
            if (typeof (obj as any)[propertyName] === "function" && !mocked[propertyName]) {
                mocked[propertyName] = jest.fn();
            }
        }
    } while (obj = Object.getPrototypeOf(obj));
    return mocked as any;
}
