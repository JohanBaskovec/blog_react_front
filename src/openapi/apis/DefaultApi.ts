// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { Observable } from 'rxjs';
import { BaseAPI, HttpHeaders, throwIfNullOrUndefined, encodeURI } from '../runtime';
import {
    Article,
    ArticleFormData,
    LoginForm,
    RegistrationForm,
    User,
} from '../models';

export interface GetArticleByIdRequest {
    id: string;
}

export interface InsertArticleRequest {
    articleFormData: ArticleFormData;
}

export interface LoginRequest {
    loginForm: LoginForm;
}

export interface RegisterRequest {
    registrationForm: RegistrationForm;
}

export interface UpdateArticleRequest {
    articleFormData: ArticleFormData;
}

/**
 * no description
 */
export class DefaultApi extends BaseAPI {

    /**
     * Get all the articles
     */
    getAllArticles = (): Observable<Array<Article>> => {
        return this.request<Array<Article>>({
            path: '/article',
            method: 'GET',
        });
    };

    /**
     * Get an article
     */
    getArticleById = ({ id }: GetArticleByIdRequest): Observable<Article> => {
        throwIfNullOrUndefined(id, 'getArticleById');

        return this.request<Article>({
            path: '/article/{id}'.replace('{id}', encodeURI(id)),
            method: 'GET',
        });
    };

    /**
     * Get the current session
     */
    getCurrentAuthenticatedUser = (): Observable<User> => {
        const headers: HttpHeaders = {
        };

        return this.request<User>({
            path: '/me',
            method: 'GET',
            headers,
        });
    };

    /**
     * Insert an article
     */
    insertArticle = ({ articleFormData }: InsertArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'insertArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'POST',
            headers,
            body: articleFormData,
        });
    };

    /**
     * Create a new HTTP session
     */
    login = ({ loginForm }: LoginRequest): Observable<User> => {
        throwIfNullOrUndefined(loginForm, 'login');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<User>({
            path: '/login',
            method: 'POST',
            headers,
            body: loginForm,
        });
    };

    /**
     * Logout
     */
    logout = (): Observable<void> => {
        const headers: HttpHeaders = {
        };

        return this.request<void>({
            path: '/logout',
            method: 'POST',
            headers,
        });
    };

    /**
     * Register
     */
    register = ({ registrationForm }: RegisterRequest): Observable<void> => {
        throwIfNullOrUndefined(registrationForm, 'register');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/register',
            method: 'POST',
            headers,
            body: registrationForm,
        });
    };

    /**
     * Update an article
     */
    updateArticle = ({ articleFormData }: UpdateArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'updateArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'PUT',
            headers,
            body: articleFormData,
        });
    };

}
