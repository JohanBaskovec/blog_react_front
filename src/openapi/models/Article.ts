// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface Article
 */
export interface Article {
    /**
     * @type {string}
     * @memberof Article
     */
    id: string;
    /**
     * @type {string}
     * @memberof Article
     */
    title: string;
    /**
     * @type {string}
     * @memberof Article
     */
    content: string;
}
