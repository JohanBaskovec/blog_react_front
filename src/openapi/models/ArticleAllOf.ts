// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import {
    User,
} from './';

/**
 * @export
 * @interface ArticleAllOf
 */
export interface ArticleAllOf {
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    title: string;
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    content: string;
    /**
     * @type {User}
     * @memberof ArticleAllOf
     */
    author: User;
    /**
     * @type {number}
     * @memberof ArticleAllOf
     */
    version: number;
}
