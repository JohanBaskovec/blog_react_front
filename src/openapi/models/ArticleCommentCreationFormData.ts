// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface ArticleCommentCreationFormData
 */
export interface ArticleCommentCreationFormData {
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    id: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    articleId: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    content: string;
}
