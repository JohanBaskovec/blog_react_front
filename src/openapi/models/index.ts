export * from './Article';
export * from './ArticleAllOf';
export * from './ArticleComment';
export * from './ArticleCommentAllOf';
export * from './ArticleCommentCreationFormData';
export * from './ArticleFormData';
export * from './Authorization';
export * from './ClientError';
export * from './HttpSession';
export * from './LoginForm';
export * from './ObjectBase';
export * from './Permission';
export * from './PermissionAuthorization';
export * from './PermissionAuthorizationAllOf';
export * from './RegistrationForm';
export * from './Role';
export * from './RoleAuthorization';
export * from './RoleAuthorizationAllOf';
export * from './ServerError';
export * from './User';
export * from './ValidationError';
export * from './WebSocketClientChatMessage';
export * from './WebSocketClientChatMessageAllOf';
export * from './WebSocketMessage';
export * from './WebSocketServerChatMessage';
export * from './WebSocketServerChatMessageAllOf';
