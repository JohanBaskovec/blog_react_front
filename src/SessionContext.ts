import React from "react";
import {User} from "./openapi/models";

export interface Session {
    user?: User;
    webSocket?: WebSocket;
}

export const SessionContext = React.createContext<Session>({
});
