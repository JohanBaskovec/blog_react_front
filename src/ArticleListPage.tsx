import React, {useContext, useEffect, useState} from "react";
import {Article} from "./openapi/models";
import {DefaultApi} from "./openapi/apis";
import "./ArticleListPage.scss";
import {AppLink} from "./AppLink";
import {SessionContext} from "./SessionContext";
import {UserService} from "./UserService";
import {ArticleList} from "./ArticleList";

export interface ArticleListPageProps {
    api: DefaultApi;
}

export function ArticleListPage({api}: ArticleListPageProps) {
    const [articles, setArticles] = useState<Article[]>([]);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<any>(null);
    useEffect(() => {
        setLoading(true);
        const getAllArticles$ = api.getAllArticles()
            .subscribe(articles => {
                    setArticles(articles);
                    setLoading(false);
                },
                (err) => {
                    setLoading(false);
                    setError(err);
                });
        return () => {
            getAllArticles$.unsubscribe();
        }
    }, []);
    const session = useContext(SessionContext);
    let body = null;

    if (loading) {
        body = <>Loading articles...</>;
    } else if (error) {
        const errorResponse = error.response ? error.response.message : "Unknown error.";
        body = <>An error happened: {errorResponse}</>;
    } else {
        body = <ArticleList articles={articles} />
    }
    return (
        <div className="ArticleListPage">
            {UserService.userHasRole('admin', session.user) ?
                <div>
                    <AppLink to="/article/new">New article</AppLink>
                </div> :
                null
            }
            {body}
        </div>
    )
}
