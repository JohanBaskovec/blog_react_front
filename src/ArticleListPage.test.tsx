import React from "react";
import {DefaultApi} from "./openapi/apis";
import {of} from "rxjs";
import {mockDefaultApi} from "./mockClass";
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import pretty from "pretty";
import {ArticleListPage} from "./ArticleListPage";
import {Article} from "./openapi/models";
import {MemoryRouter} from "react-router-dom";
import {Session, SessionContext} from "./SessionContext";

let container: HTMLDivElement;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    if (container != null) {
        unmountComponentAtNode(container);
        container.remove();
    }
});

describe('ArticleList', () => {
    it("renders the article list when there are 3 articles and user is logged in", () => {
        const api: DefaultApi = mockApiWith3Articles();
        const session: Session = mockLoggedInSession();
        act(() => {
            render(
                <MemoryRouter initialEntries={['/']}>
                    <SessionContext.Provider value={session}>
                        <ArticleListPage api={api}/>
                    </SessionContext.Provider>
                </MemoryRouter>
                , container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });

    it("renders the article list when there are 3 articles and user is logged out", () => {
        const api = mockApiWith3Articles();
        act(() => {
            render(
                <MemoryRouter initialEntries={['/']}>
                    <ArticleListPage api={api}/>
                </MemoryRouter>
                , container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });

    function mockApiWith3Articles() {
        const api: DefaultApi = mockDefaultApi(new DefaultApi());
        api.getAllArticles = jest.fn(() => {
            const articles: Array<Article> = [
                {
                    id: "id0",
                    title: "title0",
                    content: "content0",
                    author: {
                        username: 'username',
                        version: 0,
                    },
                    version: 0,
                    creationTime: +new Date(),
                },
                {
                    id: "id1",
                    title: "title1",
                    content: "content1",
                    author: {
                        username: 'username',
                        version: 0,
                    },
                    version: 0,
                    creationTime: +new Date(),
                },
                {
                    id: "id2",
                    title: "title2",
                    content: "content2",
                    author: {
                        username: 'username',
                        version: 0,
                    },
                    version: 0,
                    creationTime: +new Date(),
                },
            ];
            return of(articles);
        });
        return api;
    }

    function mockLoggedInSession(): Session {
        return {
            user: {
                password: "",
                username: "",
                version: 0,
            }
        };
    }
});
