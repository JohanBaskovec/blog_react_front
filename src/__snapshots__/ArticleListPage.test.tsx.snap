// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList renders the article list when there are 3 articles and user is logged in 1`] = `
"<div class=\\"ArticleListPage\\">
  <div class=\\"ArticleList\\">
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id0\\">title0</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content0</div>
    </div>
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id1\\">title1</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content1</div>
    </div>
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id2\\">title2</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content2</div>
    </div>
  </div>
</div>"
`;

exports[`ArticleList renders the article list when there are 3 articles and user is logged out 1`] = `
"<div class=\\"ArticleListPage\\">
  <div class=\\"ArticleList\\">
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id0\\">title0</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content0</div>
    </div>
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id1\\">title1</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content1</div>
    </div>
    <div class=\\"ArticleSummary\\">
      <div class=\\"ArticleSummary__title\\"><a class=\\"AppLink AppLink--blue undefined\\" href=\\"/article/id2\\">title2</a></div>
      <div class=\\"ArticleSummary__publication_time\\">publication_date Dec 29, 2020, 9:52 PM</div>
      <div class=\\"ArticleSummary__author\\">published_by <a class=\\"AppLink AppLink--blue undefined\\" href=\\"/profile/username\\">username</a></div>
      <div class=\\"ArticleSummary__content\\">content2</div>
    </div>
  </div>
</div>"
`;
